lib/system/mm/malloc.nim

```diff
@@ -3,26 +3,53 @@
 
 proc allocImpl(size: Natural): pointer =
   result = c_malloc(size.csize_t)
-  when defined(zephyr):
+  when defined(zephyr) or defined(handleOOM):
     if result == nil:
       raiseOutOfMem()
 
 proc alloc0Impl(size: Natural): pointer =
   result = c_calloc(size.csize_t, 1)
-  when defined(zephyr):
+  when defined(zephyr) or defined(handleOOM):
     if result == nil:
       raiseOutOfMem()
 
 proc reallocImpl(p: pointer, newSize: Natural): pointer =
   result = c_realloc(p, newSize.csize_t)
-  when defined(zephyr):
+  when defined(zephyr) or defined(handleOOM):
     if result == nil:
       raiseOutOfMem()
 
+when defined(handleOOM):
+  proc nRealloc*(p: pointer, oldsize: csize_t, newsize: csize_t): pointer {.inline.} =
+    result = c_calloc(newSize.csize_t, sizeof(cchar).csize_t)
+    if result == nil:
+      raiseOutOfMem()
+    else:
+      if p != nil and newsize != 0 and newsize >= 0:
+        for i in 0..(newsize - 1):
+          if i <= oldSize:
+            cast[ptr UncheckedArray[cchar]](result)[i] = cast[ptr UncheckedArray[cchar]](p)[i]
+          else:
+            cast[ptr UncheckedArray[cchar]](result)[i] = '\0'
+        c_free(p)
+    else:
+      raiseOutOfMem()
+
+
+
 proc realloc0Impl(p: pointer, oldsize, newSize: Natural): pointer =
-  result = realloc(p, newSize.csize_t)
-  if newSize > oldSize:
-    zeroMem(cast[pointer](cast[uint](result) + uint(oldSize)), newSize - oldSize)
+  when defined(handleOOM):
+    result = nRealloc(p, oldsize.csize_t, newSize.csize_t)
```

Comment:
Ha, good one. As if you could use c_calloc in Nim's allocator and get away with it.